### PR TITLE
EPMLSTRCMW-283 refactor project versioning

### PR DESCRIPTION
**Pull request by myazinn**

### CODE REVIEW

File: services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala [hunk 1]
```diff
@@ -235,36 +215,45 @@ class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
...
-    val fileVersion: String = version match {
-      case Some(version) => version.name
-      case None          => getDefaultProjectVersion().toString
+  private def getUpdatedProjectVersion(
+    project: Project,
+    optionUserVersion: Option[PipelineVersion]
+  ): AsyncResult[PipelineVersion] =
+    getLastProjectVersion(project).map(projectVersion => getNewProjectVersion(projectVersion, optionUserVersion))
+
+  private def getLastProjectVersion(project: Project): Future[Option[PipelineVersion]] =
```
Comment (Nikolai-Isaev) [inline services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:224]: Quite confusing for me. Is project.version is the same version type that must be returned? If so, why cant we use project.version? How can it be, that passed project has older version. 
If they are different entities, the naming is not very clear.

File: services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala [hunk 2]
```diff
@@ -1,58 +1,91 @@
...
+   */
+  def createRepository(localProject: LocalProject): AsyncResult[Project]
 
-  def getProjectVersions(project: Project)(
-    implicit ec: ExecutionContext
-  ): AsyncResult[Seq[GitLabVersion]]
+  /**
+   * Update files in repository with versioning
+   * @param project project to which file belongs to
+   * @param projectFile files to update
```
Comment (Nikolai-Isaev) [inline services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala:23]: Maybe file should be single here

Comment (Nikolai-Isaev) [inline services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala]: Maybe file should be single here

File: services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala [hunk 3]
```diff
@@ -1,58 +1,91 @@
...
+    project: Project,
+    projectFiles: ProjectFiles,
+    version: Option[PipelineVersion]
+  ): AsyncResult[PipelineVersion]
 
-  def getFileVersions(project: Project, path: Path)(
-    implicit ec: ExecutionContext
-  ): AsyncResult[Seq[GitLabVersion]]
+  /**
+   * Retrieves file content from repository
```
Comment (Nikolai-Isaev) [inline services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala]: Since `ProjectFileContent` is a parameter of `ProjectFile` and the method returns a whole ProjectFile, this may be confusing

File: services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala (hunk 1 above)
Comment (Nikolai-Isaev) [inline services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:224]: Can't grasp, how project.version differs from the returned version. How they relate? Can `project` contain outdated version? If they are not related, maybe something with the naming can be done.

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala [hunk 4]
```diff
@@ -61,16 +63,12 @@ object ProjectFileService {
       projectFile: ProjectFile,
       version: Option[PipelineVersion],
       userId: UserId
-    ): Future[Either[VersioningException, UpdateFiledResponse]] =
+    ): Future[Either[VersioningException, Unit]] =
```
Comment (Nikolai-Isaev) [inline services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala:58]: uploadFile in ProjectFileController still has `uploadResponse` as variable name, if that is even an issue

File: services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala (hunk 2 above)
Comment (myazinn) [reply services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala]: Agree, fixed

File: services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala (hunk 3 above)
Comment (myazinn) [reply services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala]: Chaned to `Retrieves project file from repository`

File: services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala (hunk 1 above)
Comment (myazinn) [reply services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:224]: I'm curious about it as well. `project.version` should have the latest version, so this method should not exist anymore.\
But changing behavior of `ProjectVersioning` is out of scope of current PR and I tried to change as little as possible.

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala (hunk 4 above)
Comment (myazinn) [reply services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala:58]: Great catch, thanks! Although I'm disapppointed that our tests didn't find it. Anyway, I've added third commit that resolves it.

**Review by DruzhininPavel — APPROVED** [review]
LGTM

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala [hunk 5]
```diff
@@ -1,9 +1,11 @@
 package cromwell.pipeline.service
 
+import java.nio.file.Path
```
Comment (Nikolai-Isaev) [inline services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala]: Were there any changes to the formatter policies? My fmt attempts to revert this change

File: services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala [hunk 6]
```diff
@@ -1,58 +1,91 @@
...
+   * @param localProject project that exists only locally and does not have real repository
+   * @return full project with versioning-specific fields (`repositoryId`, `version`)
+   */
+  def createRepository(localProject: LocalProject): AsyncResult[Project]
 
-  def getProjectVersions(project: Project)(
-    implicit ec: ExecutionContext
-  ): AsyncResult[Seq[GitLabVersion]]
+  /**
+   * Update files in repository with versioning
```
Comment (Nikolai-Isaev) [inline services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala]:
```suggestion
   * Update file in repository with versioning
```

File: services/src/test/resources/application.conf [hunk 7]
```diff
@@ -2,7 +2,7 @@
       gitlab {
         url = "https://gitlab.com/api/v4/"
         token = "XXXX" //TODO: Change for real GitLab private token
-        defaultFileVersion = "v.0.0.1"
+        defaultFileVersion = "v0.0.1"
         defaultBranch = "master"
       }
   }
```
Comment (Nikolai-Isaev) [inline services/src/test/resources/application.conf]: Checker is grumpy about absence of a new line at the end of file

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala (hunk 5 above)
Comment (myazinn) [reply services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala]: I don't think we have such settings and don't know why Intellij decided to change it :) Rolled back to how it was before.

File: services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala (hunk 6 above)
Comment (myazinn) [reply services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala]: Done

File: services/src/test/resources/application.conf (hunk 7 above)
Comment (myazinn) [reply services/src/test/resources/application.conf]: Fixed

**Review by Nikolai-Isaev — APPROVED** [review]
Great!

**Review by gominasian — APPROVED** [review]
LGTM

**Review by GrigoriiBerezin — APPROVED** [review]
Nice
